keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -323,6 +335,146 @@ myOtherField: newly added field
+These fields are not ratcheted because changing map keys or list type to a map alters 
+the structure of the schema, and old values cannot be correlated to new. 
+
+Additionally Server-Side-Apply does not tolerate updates that violate the map 
+keys constraint: these properties are unsafe to change without bumping CRD 
+version.
+
+### Atomic Lists and Maps
+
+Following Kubernetes convention that fields annotated `x-kubernetes-list-type`
+or `x-kubernetes-map-type` as `atomic` change as a unit, the ratcheting rules
+do not apply to values nested inside atomic lists and maps. But a DeepEqual 
+comparison on the entire object can still ratchet errors for subfields.
+
+### Weaker Ratcheting Rule for Value Validations
+
+Above we propose a conservative catch-all ratcheting rule that should provide
+a minimum level of ratcheting support for any validation rule in the schema.
+
+`DeepEqual(OldObject, NewObject)` requirement is very strong. But for some 
+validation rules it is too strict. Consider the `required` error. This
+validation is applied to the object owning the property. Thus, if a new field
+becomes required the object cannot have other fields added onto it or changed
+without also adding the newly required field.
+
+Kubernetes has precedent for using a weaker form of racheting when validation is
+strengthened:
+
+```go
+if !Validation(old) {
+  // Ignore errors caused by !Validation(new)
+}
+```
+
+An example of this is the structural schema requirement added when Kubernetes
+went GA ([source](https://github.com/kubernetes/kubernetes/blob/12dc19d46fb97cbbfeb1e12b8a10ff7ae73d9515/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L1583-L1590)): 
+```go
+// requireStructuralSchema returns true if schemas specified must be structural
+func requireStructuralSchema(oldCRDSpec *apiextensions.CustomResourceDefinitionSpec) bool {
+	if oldCRDSpec != nil && specHasNonStructuralSchema(oldCRDSpec) {
+		// don't tighten validation on existing persisted data
+		return false
+	}
+	return true
+}
+```
+
+The following fields are all the value validations allowed in our schemas:
+
+- Format
+- Maximum
+- ExclusiveMaximum
+- Minimum
+- ExclusiveMinimum
+- MaxLength
+- MinLength
+- Pattern
+- MultipleOf
+- Enum
+- MaxItems
+- MinItems
+- UniqueItems
+- MaxProperties
+- MinProperties
+- Required
+- XValidations (CEL rules, excluding rules using oldSelf)
+
+The following are structural validations also allowed to be ratcheted: 
+- Nullable
+- Properties
+
+We will apply the weaker validation rule whenever these are applied: these 
+validations will only apply if they had passed for the old value.
+
+For example:
+```go
+func validateFormat(new, old interface{}, typ, format string) *field.Error {
+  if MatchesFormat(typ, format, old) && !MatchesFormat(typ, format, new) {
+    return field.Invalid(...)
+  }
+  return nil
+}
+```
+
+#### Properties
+
+Unknown properties will be allowed in an update object for separable maps
```

Comment:
Can they change though?

Reply:
After offline discussion, not allowing this